source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -328,6 +352,52 @@ void Filter::setEncoderFilterCallbacks(Http::StreamEncoderFilterCallbacks& callb
   watermark_callbacks_.setEncoderFilterCallbacks(&callbacks);
 }
 
+void Filter::sendRequest(ProcessingRequest&& req, bool end_stream) {
+  if (config_->grpcService().has_value()) {
```

Comment:
It looks like you trying to replicate C++ polymorphism here. Is ti possible to have one interface common to both gRPC and HTTP, so the `if` statement is not required?

Reply:
This will incur code refactory to move the stream_  variable https://github.com/envoyproxy/envoy/blob/634f71e38b706b10d3591098dbda5eb65fc34097/source/extensions/filters/http/ext_proc/ext_proc.h#L484  from Filter object into the ExternalProcessorClientImpl object.  I would prefer not do this while we installing the basic functionality of HTTP in ext_proc.

Reply:
I added this as a TBD item in: https://github.com/envoyproxy/envoy/issues/35488

Reply:
Why is it so big? Using dynamic_cast like so is an antipattern and you should not really need it. Looking at the API here you just need sendRequest that takes both end_stream and stream_id.

Reply:
The issue is for gRPC client, the client->sendRequest() eventually has to call:
stream_->send(std::move(req), end_stream);

And the stream_ is not available in class ExternalProcessorClient. Refactor it from Filter class and move it to this client will trigger some work as stream_ is is in a lot of places in the code.

Reply:
Created https://github.com/envoyproxy/envoy/pull/36228 to refactor stream_ out from Filter and moving it into client_.

Reply:
Done of this.

I ends up passing a stream interface to the client_->sendRequest() API instead of storing the stream_ ptr in the client object.  Thus part of https://github.com/envoyproxy/envoy/pull/36228 is reworked, and stream_ ptr is put back to Filter class to make things simple.